service/storage_service.cc

```diff
@@ -5348,6 +5349,95 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+
+    // Check if the request is still valid.
+    // If there is mismatch, it means this streaming was canceled and the coordinator moved on.
+    if (!trinfo) {
+        throw std::runtime_error(format("No transition info for tablet {}", tablet));
+    }
+    if (trinfo->stage != locator::tablet_transition_stage::streaming) {
+        throw std::runtime_error(format("Tablet {} stage is not at streaming", tablet));
+    }
+    if (trinfo->pending_replica.host != tm->get_my_id()) {
+        throw std::runtime_error(format("Tablet {} has pending replica different than this one", tablet));
+    }
+
+    auto& tinfo = tmap.get_tablet_info(tablet.tablet);
+    auto range = tmap.get_token_range(tablet.tablet);
+    locator::tablet_replica leaving_replica = get_leaving_replica(tinfo, *trinfo);
+    if (leaving_replica.host == tm->get_my_id()) {
+        // The algorithm doesn't work with tablet migration within the same node because
+        // it assumes there is only one tablet replica, picked by the sharder, on local node.
+        throw std::runtime_error(format("Cannot stream within the same node, tablet: {}, shard {} -> {}",
+                                        tablet, leaving_replica.shard, trinfo->pending_replica.shard));
+    }
+    auto leaving_replica_ip = host2ip(leaving_replica.host);
+
+    auto& table = _db.local().find_column_family(tablet.table);
+    std::vector<sstring> tables = {table.schema()->cf_name()};
+    auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, tm, _abort_source,
+           get_broadcast_address(), _sys_ks.local().local_dc_rack(),
+           "Tablet migration", streaming::stream_reason::tablet, std::move(tables));
+    streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
+            _gossiper.get_unreachable_members()));
+
+    std::unordered_map<inet_address, dht::token_range_vector> ranges_per_endpoint;
+    ranges_per_endpoint[leaving_replica_ip].emplace_back(range);
+    streamer->add_rx_ranges(table.schema()->ks_name(), std::move(ranges_per_endpoint));
+
+    if (!_tablet_streaming[tablet]) {
+        _tablet_streaming[tablet] = with_gate(_async_gate, [this, leaving_replica, tablet, &streamer] () mutable -> future<> {
+            auto streamer_ = std::move(streamer);
+            auto leaving_replica_ = leaving_replica;
+            auto tablet_ = tablet;
+            try {
+                co_await streamer_->stream_async();
+                slogger.info("Streaming for tablet migration of {} successful", tablet_);
+                _tablet_streaming.erase(tablet_);
+            } catch (...) {
+                _tablet_streaming.erase(tablet_);
+                auto ep = std::current_exception();
+                slogger.warn("Streaming for tablet migration of {} from {} failed: {}", tablet_, leaving_replica_, ep);
+                std::rethrow_exception(std::move(ep));
```

Comment:
"throw;" works just as well, and then you don't need "ep".

Reply:
Right, I just copied it from somewhere